sound/soc/sof/ipc3.c

```diff
@@ -898,6 +898,35 @@ static void ipc3_xrun(struct snd_sof_dev *sdev, u32 msg_id)
 #endif
 }
 
+static void ipc3_drain(struct snd_sof_dev *sdev, u32 msg_id)
+{
+	struct snd_soc_component *scomp = sdev->component;
+	struct snd_sof_pcm *spcm;
+	struct snd_sof_pcm_stream *stream;
+	struct sof_ipc_stream_drain drain;
+	int direction;
+
+	spcm = snd_sof_find_spcm_comp(scomp, msg_id, &direction);
+	if (!spcm) {
+		dev_err(sdev->dev,
+			"error: drain for unknown stream, msg_id %d\n",
+			msg_id);
+		return;
+	}
+
+	stream = &spcm->stream[direction];
+	snd_sof_ipc_msg_data(sdev, stream->substream, &drain, sizeof(drain));
+
+	/* since struct sof_ipc_stream drain only contains the reply
+	 * header there's no need for additional checks.
+	 *
+	 * we're assuming that since we're receiving the drain-related
+	 * reply then the drain operation finished successfully.
+	 */
+	if (spcm->pcm.compress)
+		snd_compr_drain_notify(stream->cstream);
```

Comment:
And what happens when the drain is finished? Don't we need a TRIGGER_STOP or something?

Reply:
Drain can be possible without stop if there are somehow two consecutive streams being fed, you do a partial drain for the first stream before starting the next one (useful for gapless playback I guess? Not 100% sure). I'd guess stop should wait until a drain finishes though.